src/actions/pkg.cr

```diff
@@ -1,44 +1,146 @@
+      raise error
+    end
+    raise NoPkgsToClean.new context
+  end
+
   # List built packages
   relative_get nil do |context|
     if context.current_user? && Config.has_access? context, Access::Read
-      # TODO: list all built packages
+      pfx = DPPM::Prefix.new prefix
+      # Build JSON directly to the HTTP response IO
+      JSON.build context.response do |json|
+        json.array do
+          # build an array of all the package names
+          pfx.each_pkg { |pkg| json.string pkg.package }
+        end
+      end
       next context
     end
     raise Unauthorized.new context
   end
   # Clean unused built packages
   relative_delete "/clean" do |context|
     if context.current_user? && Config.has_access? context, Access::Delete
-      # TODO: delete all unused built packages
+      clean_unused_packages(context).to_json context.response
+      context.response.flush
       next context
     end
     raise Unauthorized.new context
   end
   # Query information about a given package
   relative_get "/:id/query" do |context|
     if context.current_user? && Config.has_access? context, Access::Read
-      # TODO: Query information about the given package
+      package_name = URI.unescape context.params.url["id"]
+      # Iterate over the packages to find the relevant one.
+      if selected_pkg = find_package_by_name package_name, prefix
+        # The package whose name was the :id URL parameter was found and we can query it
+        if key = context.params.query["get"]?
+          # A key was specified by the &get query parameter
+          begin
+            {data: {
+              package_name => {
+                key => selected_pkg.get_config key,
+                # This method call  ^^ raises an untyped error if the key is not found
+              },
+            }}.to_json context.response
+            context.response.flush
+            next context
+          rescue e : Exception
+            # catch the exception thown by the lack of a config value for the specified key
+            # otherwise, pass it on to the
+            msg = e.message || raise e
+            raise e unless msg.starts_with? "config key not found"
+            context.response.status_code = 404
+            {errors: [e.message]}.to_json context.response
+            context.response.flush
+            next context
+          end
+        else
+          # No specific key was requested, respond with all of the config options
+          # for this package and all dependent packages
+          JSON.build context.response do |builder|
+            builder.object do
+              output_config_for selected_pkg, to: builder
+              if context.params.query["deps"]? && (deps = selected_pkg.pkg_file.deps)
+                builder.field "dependencies" do
+                  deps.each do |name, version|
+                    semver = SemanticVersion.parse version
+                    DPPM::Prefix.new(prefix).each_pkg do |pkg|
```

Comment:
Why iterating over all packages?

Reply:
"libraries" or "installed libraries" is a more correct word, because there are optional dependencies like `git` (for the filebrowser application).
You can use `Pkg#libs`.